contracts/Enforcer.sol

```diff
@@ -1,134 +1,184 @@
     }
 
-    /**
-      * @dev dispute is called by challenger to start a new dispute
-      *     assumed that challenger's execution tree is of the same depth as solver's
-      *     in case challenger's tree is shallower, he should use node with zero hash to make it deeper
-      *     in case challenger's tree is deeper, he should submit only the left subtree with the same depth with solver's
-      */
-    function dispute(address codeContractAddress, bytes memory _callData, bytes32 endHash)
+    function dispute(bytes32 _solverPathRoot, bytes32 _challengerPathRoot, EVMParameters memory _parameters)
         public payable
     {
-        bytes32 executionId = keccak256(abi.encodePacked(codeContractAddress, _callData));
+        bytes32 taskHash = parameterHash(_parameters);
+        bytes32 executionId = keccak256(abi.encodePacked(taskHash, _solverPathRoot));
+        ExecutionResult memory executionResult = executions[executionId];
 
-        Execution storage execution = executions[executionId];
-
-        require(execution.startBlock != 0, "Execution does not existed");
-        // executionDepth round plus 1 for submitProof
-        require(execution.startBlock + challengePeriod > block.number + (execution.executionDepth + 1) * verifier.timeoutDuration(),
-                "Execution is out of challenge period");
         require(msg.value == bondAmount, "Bond amount is required");
-        // Do we want to prohibit early?
-        // require(execution.endHash != endHash);
+        require(executionResult.startBlock != 0, "Execution does not exist");
+        // executionDepth round plus 1 for submitProof
+        require(
+            executionResult.startBlock + challengePeriod > block.number + (executionResult.executionDepth + 1) * verifier.timeoutDuration(),
+            "Execution is out of challenge period"
+        );
 
         bonds[msg.sender] += bondAmount;
 
+        // TODO: Verifier needs to support all EVMParameters
         bytes32 disputeId = verifier.initGame(
             executionId,
-            execution.endHash,
-            endHash,
-            execution.executionDepth,
-            execution.customEnvironmentHash,
+            _solverPathRoot,
+            _challengerPathRoot,
+            executionResult.executionDepth,
+            _parameters.customEnvironmentHash,
+            _parameters.codeHash,
+            _parameters.dataHash,
             // challenger
-            msg.sender,
-            codeContractAddress,
-            _callData
+            msg.sender
         );
 
         emit DisputeInitialised(disputeId, executionId);
     }
 
-    // receive result from Verifier contract
-    function result(bytes32 executionId, bool solverWon, address challenger) public onlyVerifier() {
-        Execution memory execution = executions[executionId];
+    /// only callable from verifier
+    function result(bytes32 _executionId, bool solverWon, address challenger) public {
```

Comment:
we agreed to handle this in a later bounty. is there an issue somewhere already that i can put requirements in?

Reply:
No Issue yet, do you want to create one or should I ?

Reply:
Here you go: https://github.com/leapdao/solEVM-enforcer/issues/140